File: src/editor/filter/leaf/options/multi/SupertypeFilter.java
package editor.filter.leaf.options.multi;

import java.util.HashSet;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

import editor.database.card.Card;
import editor.filter.Filter;
import editor.filter.FilterAttribute;

/**
 * This class represents a filter that groups cards by supertype.
 *
 * @author Alec Roelke
 */
public class SupertypeFilter extends MultiOptionsFilter<String>
{
    /**
     * List of all supertypes that appear on cards.
     */
    public static String[] supertypeList = {};

    /**
     * Create a new SupertypeFilter.
     */
    public SupertypeFilter()
    {
        super(FilterAttribute.SUPERTYPE, Card::supertypes);
    }

    @Override
    protected String convertFromString(String str)
    {
        return str;
    }

    @Override
    public Filter copy()
    {
        SupertypeFilter filter = (SupertypeFilter)FilterAttribute.createFilter(FilterAttribute.SUPERTYPE);
        filter.contain = contain;
        filter.selected = new HashSet<>(selected);
        return filter;
    }

    @Override
    protected JsonElement convertToJson(String item)
    {
        return new JsonPrimitive(item);
    }
}

File: src/main/scala/editor/filter/leaf/options/multi/SupertypeFilter.scala
package editor.filter.leaf.options.multi

import editor.database.attributes.CardAttribute
import editor.filter.FaceSearchOptions
import editor.util.Containment

/**
 * Object containing global information about supertypes.
 * @author Alec Roelke
 */
object SupertypeFilter {
  /** List of all supertypes. */
  var supertypeList = Seq[String]()
}

/**
 * Filter that groups cards by supertype.
 * @author Alec Roelke
 */
final case class SupertypeFilter(faces: FaceSearchOptions = FaceSearchOptions.ANY, contain: Containment = Containment.AnyOf, selected: Set[String] = Set.empty) extends MultiOptionsFilter[String, SupertypeFilter] {
  override def attribute = CardAttribute.Supertype
  override val unified = false
  override val values = _.supertypes
  override def copyFaces(faces: FaceSearchOptions) = copy(faces = faces)
}